data/dataset.py

```diff
@@ -0,0 +1,144 @@
+import torch
+import math
+from torch_geometric.data import Data, HeteroData, InMemoryDataset
+from torch import Tensor
+from typing import Union, Optional
+
+
+def get_negative_edges(
+    filter_ids: Tensor,
+    potential_edges: Tensor,
+    num_negative_edges: int = 10,
+) -> Tensor:
+    # Get the biggest value available in articles (potential edges to sample from)
+    id_max = torch.max(potential_edges, dim=1)[0][1]
+
+    # Create list of potential negative edges, filter out positive edges
+    combined = torch.cat((torch.range(start=0, end=id_max), filter_ids))
+    uniques, counts = combined.unique(return_counts=True)
+    difference = uniques[counts == 1]
+
+    # Randomly sample negative edges
+    negative_edges = difference[torch.randperm(difference.nelement())][
+        :num_negative_edges
+    ]
+
+    return negative_edges
+
+
+def remap_indexes_to_zero(
+    all_edges: Tensor, buckets: Optional[Tensor] = None
+) -> Tensor:
+    all_edges_copy = all_edges.clone()
```

Comment:
is this necessary? does bucketize mutate the input tensor? my naive assumption is that it won't, as it returns something else

Reply:
It's not necessary anymore (I was using a different method before that was mutating it)